worker/remoterelations/remoterelations.go

```diff
@@ -339,12 +345,48 @@ func (w *remoteApplicationWorker) loop() error {
 	}
 }
 
-func (w *remoteApplicationWorker) killRelationUnitWatcher(key string, relations map[string]*relation) error {
+func (w *remoteApplicationWorker) processRelationGone(
+	key string, relations map[string]*relation, publisher RemoteRelationChangePublisher,
+) error {
+	logger.Debugf("relation %v gone", key)
 	relation, ok := relations[key]
-	if ok {
-		delete(relations, key)
-		return worker.Stop(relation.ruw)
+	if !ok {
+		return nil
+	}
+	delete(relations, key)
+	if err := worker.Stop(relation.ruw); err != nil {
+		logger.Warningf("stopping relation unit watcher for %v: %v", key, err)
+	}
+
+	// Remove the remote entity record for the relation to ensure any unregister
+	// call from the remote model that may come across at the same time is short circuited.
+	remoteId := relation.ruw.remoteRelationId
+	_, err := w.facade.GetToken(remoteId.ModelUUID, names.NewRelationTag(key))
+	if errors.IsNotFound(err) {
+		logger.Debugf("not found token for %v in %v, exit early", key, w.localModelUUID)
+		return nil
+	} else if err != nil {
+		return errors.Trace(err)
 	}
+
+	// We also need to remove the remote entity reference for the relation.
+	if err := w.facade.RemoveRemoteEntity(remoteId.ModelUUID, names.NewRelationTag(key)); err != nil {
+		return errors.Trace(err)
+	}
+
+	// We need to tell the other model that the relation is gone.
+	errResults, err := publisher.UnregisterRemoteRelations(remoteId)
```

Comment:
So it sounds like the process here is that we remove the relation entity from the local model, then we call unregister on the remote model. And then that would trigger the same worker in the other worker, which wakes up, tries to follow the same process but stops because it doesn't find the relation entity in this model in its own call to UnregisterRemoteRelations? Is that vaguely right?

It seems like this is a bit asymmetrical - why does one or the other worker do all the work? Or rather, it seems like all the interesting work is done in UnregisterRemoteRelation which gets called on the other model. Couldn't most of the work be done locally, with just one cross-model call that only does enough to trigger the other worker?